### PR TITLE
Extended codec

### PR DESCRIPTION
**Pull request by StephenButtolph**

### CODE REVIEW

File: codec/reflectcodec/struct_fielder.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+func NewStructFielder(tagName string, maxSliceLen int) StructFielder {
+	return &structFielder{
+		tagName:                tagName,
+		maxSliceLen:            maxSliceLen,
+		serializedFieldIndices: make(map[reflect.Type][]FieldDesc),
+	}
+}
+
+// Codec handles marshaling and unmarshaling of structs
```
Comment (danlaine) [inline codec/reflectcodec/struct_fielder.go:39]: Copy pasta error

File: codec/reflectcodec/type_codec.go [hunk 2]
```diff
@@ -31,69 +24,25 @@ var (
...
-	// Key: a struct type
-	// Value: Slice where each element is index in the struct type
-	// of a field that is serialized/deserialized
-	// e.g. Foo --> [1,5,8] means Foo.Field(1), etc. are to be serialized/deserialized
-	// We assume this cache is pretty small (a few hundred keys at most)
-	// and doesn't take up much memory
-	serializedFieldIndices map[reflect.Type][]int
+type TypeCodec interface {
+	UnpackPrefix(*wrappers.Packer, reflect.Type) (reflect.Value, error)
+	PackPrefix(*wrappers.Packer, reflect.Type) error
```
Comment (danlaine) [inline codec/reflectcodec/type_codec.go:29]: // PackPrefix packs the prefix for the given type into the given packer.
// This identifies the bytes that follow, which are the byte representation of an interface, as having the given concrete type.
// When deserializing the bytes, the prefix specifies which concrete type to deserialize into.

File: codec/reflectcodec/type_codec.go [hunk 3]
```diff
@@ -31,69 +24,25 @@ var (
...
-	fieldLock sync.Mutex
-	// Key: a struct type
-	// Value: Slice where each element is index in the struct type
-	// of a field that is serialized/deserialized
-	// e.g. Foo --> [1,5,8] means Foo.Field(1), etc. are to be serialized/deserialized
-	// We assume this cache is pretty small (a few hundred keys at most)
-	// and doesn't take up much memory
-	serializedFieldIndices map[reflect.Type][]int
+type TypeCodec interface {
+	UnpackPrefix(*wrappers.Packer, reflect.Type) (reflect.Value, error)
```
Comment (danlaine) [inline codec/reflectcodec/type_codec.go:28]: // UnpackPrefix unpacks the prefix of an interface from the given packer.
// The prefix specifies the concrete type that the interface should be deserialized into.
// This function returns a new instance of that concrete type.
// The concrete type must implement the given type.

File: codec/reflectcodec/type_codec.go [hunk 4]
```diff
@@ -110,21 +59,18 @@ func (c *codec) RegisterType(val interface{}) error {
 // 8) nil slices are marshaled as empty slices
 
 // To marshal an interface, [value] must be a pointer to the interface
-func (c *codec) MarshalInto(value interface{}, p *wrappers.Packer) error {
+func (c *genericCodec) MarshalInto(value interface{}, p *wrappers.Packer) error {
```
Comment (danlaine) [inline codec/reflectcodec/type_codec.go:72]: Above, we should remove `// 1) See codec_test.go for examples of usage`

File: codec/reflectcodec/struct_fielder.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+)
+
+type FieldDesc struct {
+	Index       int
+	MaxSliceLen int
+}
+
+// StructFielder handles discovery of serializable fields in a struct.
+type StructFielder interface {
+	GetSerializedFields(t reflect.Type) ([]FieldDesc, error)
```
Comment (danlaine) [inline codec/reflectcodec/struct_fielder.go:29]:
```
// Returns the fields that have been marked as serializable in [t], which is a struct type.
// Additionally, returns the custom maximum length slice that may be serialized into the field, if any.
//  Returns an error if a field has tag "[tagName]: [TagValue]" but the field is un-exported. 
// In addition to
// e.g. GetSerializedField(Foo) --> [1,5,8] means
// Foo.Field(1), Foo.Field(5), Foo.Field(8) are to be serialized/deserialized
```

File: codec/linearcodec/codec.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	return nil
+}
+
+func (c *linearCodec) PackPrefix(p *wrappers.Packer, valueType reflect.Type) error {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	typeID, ok := c.typeToTypeID[valueType] // Get the type ID of the value being marshaled
+	if !ok {
+		return fmt.Errorf("can't marshal unregistered type '%v'", valueType.String())
```
Comment (danlaine) [inline codec/linearcodec/codec.go:82]: Is there a reason not to do this instead?
`return fmt.Errorf("can't marshal unregistered type '%s'", valueType)`

File: codec/reflectcodec/type_codec.go [hunk 7]
```diff
@@ -31,100 +24,63 @@ var (
...
+// 2) To include a field of a struct in the serialized form, add the tag `{tagName}:"true"` to it. `{tagName}` defaults to `serialize`.
+// 3) These typed members of a struct may be serialized:
+//    bool, string, uint[8,16,32,64], int[8,16,32,64],
+//	  structs, slices, arrays, interface.
+//	  structs, slices and arrays can only be serialized if their constituent values can be.
+// 4) To marshal an interface, you must pass a pointer to the value
+// 5) To unmarshal an interface,  you must call codec.RegisterType([instance of the type that fulfills the interface]).
+// 6) Serialized fields must be exported
+// 7) nil slices are marshaled as empty slices
+type genericCodec struct {
```
Comment (danlaine) [inline codec/reflectcodec/type_codec.go:56]: Can we add a comment that this struct isn't safe for concurrent use?

Comment (StephenButtolph) [reply codec/reflectcodec/type_codec.go:56]: What's not safe about it?

Comment (danlaine) [reply codec/reflectcodec/type_codec.go:56]: nvm, I forgot we moved the lock to the "struct fielder"

**Review by danlaine — APPROVED** [review]

**Pull request merged**